cpp/src/arrow/compute/exec/expression.cc

```diff
@@ -107,9 +107,10 @@ ValueDescr Expression::descr() const {
   return CallNotNull(*this)->descr;
 }
 
+const std::shared_ptr<DataType> kNoType;
```

Comment:
Nit: wrap it inside an anonymous namespace, or as static global?

Did a quick comparison of `local static` and `global`. Looks for normal case (not the first time access which is complex due to thread safety), `local static` has the penalty of additional load-compare-cond_jump. Guess it may cause observable effect on microbenchmark.
https://godbolt.org/z/cxWPhccnx